### DISCUSSION TITLE
Support queryKey on hooks those rely on useQuery

### DISCUSSION THREAD

**Discussion by ArianHamdi** [wevm/wagmi#2739] [General]
Currently wagmi generate `queryKey` automatically for each custom hook that rely on `useQuery`, it simplies use of wagmi custom hook but it limit some react-query benefits.

in some scenario I need to manually invalidate some queries by `queryClient.invalidateQueries` or update a query cache by calling `queryClient.setQueryData` but as I don't know the exact `queryKey` for those custom hook, this functionaly are became unusable.

I had to migrate to wagmi/core from wagmi/react and use wagmi core functionality and manually manage react-query to add custom `queryKey` to the `useQuery` to access them in react-query functions.

I like to use wagmi/react but I need to specify `queryKey` for each custom hook.

We could do it two ways:
First, we should allow the user to define a `queryKey`. If a `queryKey` is provided, the hook will use it. Otherwise, wagmi will automatically generate one. 

Second, allow user to define a `queryKey` and then wagmi add this `queryKey` to the begining of auto generated `queryKey`. this is something like `scopeKey` but at the begining.

In this way when something like `account` or `chainId` are changed the query will refetch because these are parts of the `queryKey` that wagmi generated, and then we can use [Filters](https://tanstack.com/query/v4/docs/react/guides/filters) to manipulate the query manually.
For ex. `queryClient.removeQueries({ queryKey: ['posts']})` will remove queries whose `queryKey` started with `posts`

**Comment by gpxl-dev** [comment]
Would also be great to be able to use the react-query devtools - at present we can't because it can't find the query provider.

**Reply by ArianHamdi** [reply]
You can create queryClient manually and pass it to both WagmiClient and RQ devtools
